src/D2L.CodeStyle.Analyzers/ApiUsage/Logging/LoggingExecutionContextScopeBuilderAnalyzer.cs

```diff
@@ -0,0 +1,218 @@
+			if( !ILoggingExecutionContextScopeBuilderRunSymbols.Any() ) {
+				return;
+			}
+
+			ImmutableHashSet<INamedTypeSymbol> taskTypeBuiltins = ImmutableHashSet
+				.Create(
+					context.Compilation.GetTypeByMetadataName( "System.Threading.Tasks.Task" ),
+					context.Compilation.GetTypeByMetadataName( "System.Threading.Tasks.Task`1" ),
+					context.Compilation.GetTypeByMetadataName( "System.Threading.Tasks.ValueTask" ),
+					context.Compilation.GetTypeByMetadataName( "System.Threading.Tasks.ValueTask`1" ),
+					context.Compilation.GetTypeByMetadataName( "System.Runtime.CompilerServices.ConfiguredTaskAwaitable" ),
+					context.Compilation.GetTypeByMetadataName( "System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1" )
+				)
+				.Where( x => x != null && x.Kind != SymbolKind.ErrorType )
+				.ToImmutableHashSet();
+
+			INamedTypeSymbol AsyncMethodBuilderAttribute = context.Compilation.GetTypeByMetadataName( "System.Runtime.CompilerServices.AsyncMethodBuilderAttribute" );
+
+			context.RegisterSyntaxNodeAction(
+				ctx => RunInvocationAnalysis(
+					context: ctx,
+					AsyncMethodBuilderAttribute: AsyncMethodBuilderAttribute,
+					ILoggingExecutionContextScopeBuilderRunSymbols: ILoggingExecutionContextScopeBuilderRunSymbols,
+					taskTypeBuiltins: taskTypeBuiltins,
+					invocationSyntax: ctx.Node as InvocationExpressionSyntax
+				),
+				SyntaxKind.InvocationExpression
+			);
+		}
+
+		private void RunInvocationAnalysis(
+			SyntaxNodeAnalysisContext context,
+			IImmutableSet<IMethodSymbol> ILoggingExecutionContextScopeBuilderRunSymbols,
+			INamedTypeSymbol AsyncMethodBuilderAttribute,
+			IImmutableSet<INamedTypeSymbol> taskTypeBuiltins,
+			InvocationExpressionSyntax invocationSyntax
+		) {
+			SemanticModel model = context.SemanticModel;
+
+			if( !IsRunInvocation(
+				model,
+				ILoggingExecutionContextScopeBuilderRunSymbols,
+				invocationSyntax,
+				context.CancellationToken
+			) ) {
+				return;
+			}
+
+			if( !TryGetActionArgument(
+				invocationSyntax,
+				out ArgumentSyntax actionArgument
+			) ) {
+				return;
+			}
+
+			ImmutableArray<ITypeSymbol> potentialReturnTypes = GetReturnTypesToCheck(
+				model,
+				actionArgument,
+				context.CancellationToken
+			);
+			foreach( ITypeSymbol returnType in potentialReturnTypes ) {
+				if( IsAwaitable(
+					taskTypeBuiltins,
+					AsyncMethodBuilderAttribute,
+					returnType
+				) ) {
+					ReportDiagnostic(
+						context: context,
+						invocationSyntax: invocationSyntax
+					);
+					return;
+				}
+			}
+		}
+
+		private static bool IsRunInvocation(
+			SemanticModel model,
+			IImmutableSet<IMethodSymbol> ILoggingExecutionContextScopeBuilderRunSymbols,
+			InvocationExpressionSyntax invocationSyntax,
+			CancellationToken ct
+		) {
+			SymbolInfo methodSymbolInfo = model.GetSymbolInfo(
+				invocationSyntax,
+				ct
+			);
+
+			if( methodSymbolInfo.Symbol != null ) {
+				if( !ILoggingExecutionContextScopeBuilderRunSymbols.Contains(

```

Comment:
suggestion/nit:

```diff
-				if( !ILoggingExecutionContextScopeBuilderRunSymbols.Contains(
-					methodSymbolInfo.Symbol.OriginalDefinition
-				) ) {
-					return false;
-				}
+				return ILoggingExecutionContextScopeBuilderRunSymbols.Contains(
+					methodSymbolInfo.Symbol.OriginalDefinition
+				);
```

and the else if/implicit else below

Reply:
I'll tweak it to not need the `else`, but I'd rather keep the separate `return`s around for easy breakpoints

Reply:
Ok